daml-script/daml3/Daml/Script/Questions/Submit.daml

```diff
@@ -10,168 +11,258 @@ module Daml.Script.Questions.Submit
 import Daml.Script.Questions.Submit.Error
-import Daml.Script.Questions.Submit.Internal (actAsNonEmpty)
 import Daml.Script.Questions.TransactionTree
 import DA.NonEmpty
+import DA.List.Total (head)
 import DA.Stack
+import DA.Functor (void)
+import DA.Bifunctor (first)
 
--- | The @submit@ command.
 data Submit = Submit with
-  actAs : NonEmpty Party
-  readAs : [Party]
-  disclosures : [Disclosure]
-  commands : [Command]
-instance IsQuestion Submit [CommandResult] where command = "Submit"
-
-submitInternal : HasCallStack => [Party] -> [Party] -> [Disclosure] -> Commands a -> Script a
-submitInternal actAs readAs disc cmds =
-  fmap cmds.continue $ lift $ Submit with
-    actAs = actAsNonEmpty actAs
-    readAs = readAs
-    disclosures = disc
-    commands = cmds.commands
-
--- | Details of the @submitMustFail@ command.
-data SubmitMustFail = SubmitMustFail with
-  actAs : NonEmpty Party
-  readAs : [Party]
-  disclosures : [Disclosure]
-  commands : [Command]
-instance IsQuestion SubmitMustFail () where command = "SubmitMustFail"
-
-submitMustFailInternal : HasCallStack => [Party] -> [Party] -> [Disclosure] -> Commands a -> Script ()
-submitMustFailInternal actAs readAs disc cmds =
-  lift $ SubmitMustFail with
-    actAs = actAsNonEmpty actAs
-    readAs = readAs
-    disclosures = disc
-    commands = cmds.commands
-
--- | Submit the commands as a single transaction.
-
--- This will error if the submission fails.
-
-#ifdef DAML_SCENARIOS
-
-instance HasSubmit Script Commands where
-  submit p cmds = submitInternal [p] [] [] cmds
-  submitMustFail p cmds = submitMustFailInternal [p] [] [] cmds
-
-#else
-
-submit : HasCallStack => Party -> Commands a -> Script a
-submit p cmds = submitInternal [p] [] [] cmds
-
-submitMustFail : HasCallStack => Party -> Commands a -> Script ()
-submitMustFail p cmds = submitMustFailInternal [p] [] [] cmds
-
-#endif
-
--- | `submitWithDisclosures p disc cmds` submits the commands `cmds` as a single transaction
--- from party `p` with the explicit disclosures `disc` and returns the value returned by `cmds`.
--- If the transaction fails, `submit` also fails.
-submitWithDisclosures : HasCallStack => Party -> [Disclosure] -> Commands a -> Script a
-submitWithDisclosures p disc cmds = submitInternal [p] [] disc cmds
+  submissions : [Submission]
+instance IsQuestion Submit [Either (Optional SubmitError) ([CommandResult], Optional TransactionTree)] where command = "Submit"
 
--- | `submitWithDisclosures p disc cmds` submits the commands `cmds` as a single transaction
--- from party `p` with the explicit disclosures `disc`
--- It only succeeds if the submitting the transaction fails.
-submitWithDisclosuresMustFail : HasCallStack => Party -> [Disclosure] -> Commands a -> Script ()
-submitWithDisclosuresMustFail p disc cmds =submitMustFailInternal [p] [] disc cmds
+data Submission = Submission with
+  sActAs : NonEmpty Party
+  sReadAs : [Party]
+  sDisclosures : [Disclosure]
+  sErrorBehaviour : ErrorBehaviour
```

Comment:
Note that we pass up the ErrorBehaviour, as for failures in `MustSucceed`, we need to throw the original scala error, not the derived daml `SubmitError`, so the runner needs this information.
On top of this, the IdeLedgerClient needs it for reporting in VSCode